r/tests/testthat/test-dplyr-string-functions.R

```diff
@@ -0,0 +1,120 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+library(dplyr)
+library(stringr)
+
+test_that("sub and gsub", {
+  df <- tibble(x = c("Foo", "bar"))
+
+  for(fun in list(quote(sub), quote(gsub))) {
```

Comment:
I like what you're trying to do here but I think the elimination of code duplication isn't worth the sacrifice in readability. Since we know that `sub` and `gsub` use the same C++ function, just with a different option, we probably don't need to test every permutation of them. Likewise with the other options. You could keep the `fixed` for loop, that one is pretty readable; the downside there is that if one of them fails, I don't think the `testthat` output will tell you which one it is (that's solvable but may require hacking into `expect_dplyr_equal`, which probably isn't worth it). 

Since the business of evaluating the regular expressions is handled (and tested) in the C++ library, our job in this test suite is to make sure that we're correctly translating the R options to the C++ options. We aren't responsible for making sure that the C++ library correctly handles all permutations of its options--we obviously want that, and if we discover ways that it doesn't handle them correctly, we should make sure that gets fixed, but the R test suite doesn't need to assert that.

Reply:
Ok, will do. I'll trim these tests so they exercise all the different paths here in the R package code but don't go beyond that.

Reply:
Improved in 7e7e748cde2c7bf698b75debecc3542f27d818bb